Initial docs (#142)
OpenZeppelin/openzeppelin-upgrades: docs/modules/ROOT/pages/writing-upgradeable.adoc

        _decimals = 18;
    }

  ...
}
----

This means you should not be using these contracts in your OpenZeppelin Upgrades project. Instead, make sure to use `@openzeppelin/contracts-ethereum-package`, which is an official fork of OpenZeppelin Contracts that has been modified to use initializers instead of constructors. Take a look at how https://github.com/OpenZeppelin/openzeppelin-contracts-ethereum-package/blob/v3.0.0/contracts/token/ERC20/ERC20.sol[ERC20] looks in `@openzeppelin/contracts-ethereum-package`:

[source,solidity]
----
// @openzeppelin/contracts-ethereum-package/contracts/token/ERC20/ERC20.sol
pragma solidity ^0.6.0;
  ...
contract ERC20UpgradeSafe is Initializable, ContextUpgradeSafe, IERC20 {
  ...
    string private _name;
    string private _symbol;
    uint8 private _decimals;

    function __ERC20_init(string memory name, string memory symbol) internal initializer {
        __Context_init_unchained();
        __ERC20_init_unchained(name, symbol);
    }

    function __ERC20_init_unchained(string memory name, string memory symbol) internal initializer {
        _name = name;
        _symbol = symbol;
        _decimals = 18;
    }
  ...
}
----

> Whether using OpenZeppelin Contracts or another Ethereum Package, always make sure that the package is set up to handle upgradeable contracts.

Reviewer: ```suggestion
Whether using OpenZeppelin Contracts or another smart contract library, always make sure that the package is set up to handle upgradeable contracts.
```

Author: Resolved in: https://github.com/OpenZeppelin/openzeppelin-upgrades/pull/142/commits/2111130ed7681ce04989a5b15a87ed4173511454

[[avoid-initial-values-in-field-declarations]]
=== Avoiding Initial Values in Field Declarations

Solidity allows defining initial values for fields when declaring them in a contract.

[source,solidity]
----
contract MyContract {
    uint256 public constant hasInitialValue = 42;
}
----

This is equivalent to setting these values in the constructor, and as such, will not work for upgradeable contracts. Make sure that all initial values are set in an initializer function as shown below; otherwise, any upgradeable instances will not have these fields set.

[source,solidity]
----
contract MyContract is Initializable {
    uint256 public hasInitialValue;

    function initialize() public initializer {
        hasInitialValue = 42;
    }
}
----

> Note that it still is fine to set constants here, because the compiler https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables[does not reserve a storage slot for these variables], and every occurrence is replaced by the respective constant expression. So the following still works with OpenZeppelin Upgrades:

Reviewer: ```suggestion
Note that it is still fine to define _constant_ state variables in this way, because the compiler https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables[does not reserve a storage slot for these variables], and every occurrence is replaced by the respective constant expression. So the following still works with OpenZeppelin Upgrades:
```

[source,solidity]
----
contract MyContract {
    uint256 public constant hasInitialValue = 42;
}
----

[[creating-new-instances-from-your-contract-code]]
== Creating New Instances From Your Contract Code

When creating a new instance of a contract from your contract's code, these creations are handled directly by Solidity and not by OpenZeppelin Upgrades, which means that *these contracts will not be upgradeable*.

For instance, in the following example, even if `MyContract` is deployed as upgradeable, the `token` contract created is not:

[source,solidity]
----
// contracts/MyContract.sol
// SPDX-License-Identifier: MIT
pragma solidity ^0.6.0;

import "@openzeppelin/upgrades/contracts/Initializable.sol";
import "@openzeppelin/contracts/token/ERC20/ERC20.sol";

contract MyContract is Initializable {
    ERC20 public token;

    function initialize() public initializer {
        token = new ERC20("Test", "TST"); // This contract will not be upgradeable
    }
}
----

> The easiest way around this issue is to simply accept an instance of that contract as a parameter, and inject it after creating it:

Reviewer: ```suggestion
If you would like the `ERC20` instance to be upgradeable, the easiest way to achieve that is to simply accept an instance of that contract as a parameter, and inject it after creating it:
```

    IERC20 public token;

    function initialize(IERC20 _token) public initializer {
        token = _token;
    }
}
----

[[potentially-unsafe-operations]]
== Potentially Unsafe Operations

When working with upgradeable smart contracts, you will always interact with the contract instance, and never with the underlying logic contract. However, nothing prevents a malicious actor from sending transactions to the logic contract directly. This does not pose a threat, since any changes to the state of the logic contracts do not affect your contract instances, as the storage of the logic contracts is never used in your project.

There is, however, an exception. If the direct call to the logic contract triggers a `selfdestruct` operation, then the logic contract will be destroyed, and all your contract instances will end up delegating all calls to an address without any code. This would effectively break all contract instances in your project.

A similar effect can be achieved if the logic contract contains a `delegatecall` operation. If the contract can be made to `delegatecall` into a malicious contract that contains a `selfdestruct`, then the calling contract will be destroyed.

> As such, it is strongly recommended to avoid any usage of either `selfdestruct` or `delegatecall` in your contracts. If you need to include them, make absolutely sure they cannot be called by an attacker on an uninitialized logic contract.

Reviewer: I think the CLI allowed force-overriding this but we don't.

```suggestion
As such, it is not allowed to use either `selfdestruct` or `delegatecall` in your contracts.
```
